play-restli/src/main/java/com/linkedin/playrestli/BaseRestliServerComponent.java

```diff
@@ -20,40 +19,34 @@
  */
 public abstract class BaseRestliServerComponent<T extends Request> {
   protected final HttpDispatcher _restliDispatcher;
+  private final ClientCookieEncoder _cookieEncoder;
 
-  protected BaseRestliServerComponent(HttpDispatcher restliDispatcher) {
+  protected BaseRestliServerComponent(HttpDispatcher restliDispatcher, CookiesConfiguration cookiesConfiguration) {
     _restliDispatcher = restliDispatcher;
-  }
-
-  /**
-   * Transform the request headers to a Map by ignoring multiple values.
-   */
-  protected Map<String, String> toSimpleMap(Map<String, List<String>> map) {
-    Map<String, String> result = new HashMap<>();
-    for (Map.Entry<String, List<String>> entry : map.entrySet()) {
-      String commaSeparatedValues = StringUtils.join(entry.getValue(), ",");
-      result.put(entry.getKey(), commaSeparatedValues);
-    }
-    return result;
+    _cookieEncoder = cookiesConfiguration.clientEncoder();
   }
 
   protected <B extends BaseRequestBuilder<B>> B createRequestBuilder(
       Http.Request request, Function<URI, B> createBuilder) throws Exception {
+    B builder = createBuilder.apply(new URI(request.uri()));
+    builder.setMethod(request.method());
 
+    for (Map.Entry<String, List<String>> header : request.getHeaders().toMap().entrySet()) {
+      String key = header.getKey();
 
-    B builder = createBuilder.apply(new URI(request.uri()));
-    Map<String, List<String>> headers = request.getHeaders().toMap();
+      if (key.equalsIgnoreCase(Http.HeaderNames.COOKIE)) {
+        continue; // Cookie header and request.cookies may be out of sync; request.cookies is the source of truth.
+      }
+
+      for (String value : header.getValue()) {
+        builder.addHeaderValue(key, value);
+      }
+    }
```

Comment:
The above could also be simplified into Java 8 style.

`request.getHeaders().toMap().entrySet().stream().filter(entry -> !entry.getKey().equalsIgnoreCase(Http.HeaderNames.COOKIE)).forEach(entry -> entry.getValue().forEach(v -> builder.addHeader(entry.getKey(), v)));`

Something like above. Might have something better.